src/main/java/org/moltimate/moltimatebackend/service/AlignmentService.java

```diff
@@ -234,4 +264,100 @@ private double rmsd(String motifId, List<Residue> activeSiteResidues, List<Group
         Point3d[] point3ds = new Point3d[points.size()];
         return points.toArray(point3ds);
     }
+
+    /**
+     * Finds all permutations of a residue mapping
+     *
+     * @param residueMapping: mapping of residue to its candidate matches for alignment
+     * @return a list of mappings from residue to group where each map is a possible alignment permutation
+     */
+    private List<Map<Residue, Group>> findAllPermutations(Map<Residue, List<Group>> residueMapping) {
+        List<Map<Residue, Group>> result = new ArrayList<>();
+        List<List<Group>> groupLists = new ArrayList<>(residueMapping.values());
+        List<List<Group>> results = cartesianProduct(groupLists);
+        List<Residue> residues = new ArrayList<>(residueMapping.keySet());
+
+        for (List<Group> res : results) {
+            Map<Residue, Group> mapping = new HashMap<>();
+            for (int i = 0; i < res.size(); i++) {
+                mapping.put(residues.get(i), res.get(i));
+            }
+            result.add(mapping);
+        }
+        return result;
+    }
+
+    /**
+     * Finds the best permutation of a list of permutations of matches to an active site
+     * of a motif
+     *
+     * @param motifStruct:  structure object of motif to find permutation of
+     * @param permutations: data structure containing your permutations
+     * @param motif:        motif that the alignment permutations relate to
+     * @return best fit permutation for alignment. We calculate this by checking if the permutation fits
+     * the constraint we have for levenstein distance and then find the one with minimum RMSD that
+     * fits this requirement.
+     */
+    private Map<Residue, Group> findBestPermutation(Structure motifStruct, List<Map<Residue, Group>> permutations, Motif motif) {
```

Comment:
I think this is still needed for when someone opts for a higher precision factor though, since it would increase the number of atom options

Reply:
I think there might be a misconception here. This has nothing to do with alt positions or chain names. What happens is that sometimes we get multiple matches for a residue.

Let's say our active site is residues [a,b]
We get a list of matches for each residue after running our query and have a map of residue to list of matches. It might look like this:
[a: [c,d] b:[e]]

So now we have a problem. Which alignment do we choose? [a:c, b:e] or [a:d, b:e]? The fact that these lists come in in an unpredictable order gives us nondeterminism if we just choose the first alignment. Sure, we can just sort the lists and then choose, but we might end up accepting a worse alignment.

What this function does is evaluate all of our options here and find the best mapping of residues for alignment. We evaluate this by checking if the levenstein distance constraint is met and then by checking the rmsd value of the superposition.

Reply:
_Spoke in person_
We can't make these any more precise because this is on the aligned structure not the motif's structure. It also does not affect that many queries, because in most cases we do not get multiple groups of matching atoms for a query set.